pkg/manifest/jsonmanifest/entry.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package jsonmanifest
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/ethersphere/bee/pkg/manifest"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// verify JSONEntry implements manifest.Entry.
+var _ manifest.Entry = (*JSONEntry)(nil)
+
+// JSONEntry is a JSON representation of a single manifest entry for a JSONManifest.
+type JSONEntry struct {
```

Comment:
Given that JSONEntry only implements manifest.Entry interface, it could be unexported and wherever it is returned or accepted as argument manifest.Entry can be used.

A different problem is with serialization and different manifest implementations. Different implementations would require compatible serialization format in order to swap them. With this approach it is expected for all implementations to have the same json encoding. In that case I would expect that both manifest.Interface and mainfest.Entry require json marshaler and unmarshaller interfaces, not arbitrary binary serialization or to define what is the serialization that is required in order to have compatibility between implementations.

Reply:
> Given that JSONEntry only implements manifest.Entry interface, it could be unexported and wherever it is returned or accepted as argument manifest.Entry can be used.

i entirely agree. in fact, i've already implemented this change in PR #503; i didn't want to add more to this PR.

> A different problem is with serialization and different manifest implementations. 

considering [the more robust version of the manifest](https://github.com/ethersphere/manifest/pull/2), we definitely need binary marshalling—but regardless of what particular implementation we use at any given moment, i think it's wise to have `manifest.Interface` embed these 2 interfaces (rather than JSON).

> Different implementations would require compatible serialization format in order to swap them. 

in a discussion with @acud and @zbiljic we thought of adding integration tests through issue #476. that is: once we pull this implementation out of `bee` and into `manifest`, we would add tests on `bee` which can verify at least _some_ things about marshaling (or other behaviors).

> With this approach it is expected for all implementations to have the same json encoding. In that case I would expect that both manifest.Interface and mainfest.Entry require json marshaler and unmarshaller interfaces, not arbitrary binary serialization or to define what is the serialization that is required in order to have compatibility between implementations.

for the `jsonmanifest.JSONEntry` type, i'm not sure whether it matters or not what type of serialization we use as long as `jsonmanifest.JSONManifest` uses the binary marshaller, but i feel like you're telling me otherwise here.

could you please explain this a bit further? specifically: why do the manifest and the entry need to use the same serialization if only the manifest interface (and not the entry) defines an encoding?

should we maybe also embed the binary marshaller signature in the `manifest.Entry` interface?

Reply:
> In that case I would expect that both manifest.Interface and mainfest.Entry require json marshaler and unmarshaller interfaces

@janos I believe that there was some confusion with serialization here. The idea was that we should not be bound to json as a serialization format. Manifests, to the API, should only define a manifest -> bytes and bytes -> manifest interface. If we coerce manifests to implement JSON marshalling then we cannot implement other encoding, whereas binary marshalling allows us to be encoding agnostic.
It is a bit confusing that we are calling `json.Marshal/Unmarshal` on the types but this is only in order to generate and parse the actual encoded data, so that we don't need to manually generate and parse the manifest json.

Reply:
Thanks for the responses @mortelli @acud. My assumption that different manifest implementation can be interchangeable with compatible serialization was wrong. Sorry. I know that binary serialization can be anything including json. Basically not all manifest implementations can be used in all places where mainfest.Interface is specified, as the serialization can differ.

The purpose of different serialization interfaces like JSON or YAML or any other specific ones is to clearly define what is the serialization possible, so that it is possible to relay on their functionalities. With having a general binary serialization interface with arbitrary serializations between implementations, it is not clear at compile time or by type checking which implementation implements which serializations. I am not sure if this was relevant for you when you designed the manifest abstraction, if not, please disregard this comment.